File: library/src/main/java/cc/shinichi/library/tool/common/HttpUtil.java
package cc.shinichi.library.tool.common;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLConnection;

/**
 * HttpURLConnection 下载图片
 */
public class HttpUtil {

    /**
     * @param urlPath     下载路径
     * @param downloadDir 下载存放目录
     * @return 返回下载文件
     */
    public static File downloadFile(String urlPath, String fileFullName, String downloadDir) {
        File file = null;
        try {
            // 统一资源
            URL url = new URL(urlPath);
            // 连接类的父类，抽象类
            URLConnection urlConnection = url.openConnection();
            // http的连接类
            HttpURLConnection httpURLConnection = (HttpURLConnection) urlConnection;
            // 设定请求的方法，默认是GET
            httpURLConnection.setRequestMethod("GET");
            // 设置字符编码
            httpURLConnection.setRequestProperty("Charset", "UTF-8");
            // 打开到此 URL 引用的资源的通信链接（如果尚未建立这样的连接）。
            httpURLConnection.connect();

            BufferedInputStream bin = new BufferedInputStream(httpURLConnection.getInputStream());

            String path = downloadDir + File.separatorChar + fileFullName;
            file = new File(path);
            if (!file.getParentFile().exists()) {
                file.getParentFile().mkdirs();
            }
            OutputStream out = new FileOutputStream(file);
            int size = 0;
            int len = 0;
            byte[] buf = new byte[1024];
            while ((size = bin.read(buf)) != -1) {
                len += size;
                out.write(buf, 0, size);
            }
            bin.close();
            out.close();
            return file;
        } catch (MalformedURLException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: library/src/main/java/cc/shinichi/library/tool/common/HttpUtil.kt
package cc.shinichi.library.tool.common

import java.io.*
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL
import java.net.URLDecoder

/**
 * HttpURLConnection 下载图片
 */
object HttpUtil {

    /**
     * @param urlPath     下载路径
     * @param downloadDir 下载存放目录
     * @return 返回下载文件
     */
    fun downloadFile(urlPath: String?, fileFullName: String, downloadDir: String): File? {
        var file: File? = null
        try {
            // 统一资源
            val url = URL(urlPath)
            // 连接类的父类，抽象类
            val urlConnection = url.openConnection()
            // http的连接类
            val httpURLConnection = urlConnection as HttpURLConnection
            // 设定请求的方法，默认是GET
            httpURLConnection.requestMethod = "GET"
            // 设置字符编码
            httpURLConnection.setRequestProperty("Charset", "UTF-8")
            // 打开到此 URL 引用的资源的通信链接（如果尚未建立这样的连接）。
            httpURLConnection.connect()
            val bin = BufferedInputStream(httpURLConnection.inputStream)
            val path = downloadDir + File.separatorChar + fileFullName
            file = File(path)
            file.parentFile?.let {
                if (!it.exists()) {
                    it.mkdirs()
                }
            }
            val out: OutputStream = FileOutputStream(file)
            var size = 0
            var len = 0
            val buf = ByteArray(1024)
            while (bin.read(buf).also { size = it } != -1) {
                len += size
                out.write(buf, 0, size)
            }
            bin.close()
            out.close()
            return file
        } catch (e: MalformedURLException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    fun decode(text: String): String {
        return URLDecoder.decode(text)
    }
}